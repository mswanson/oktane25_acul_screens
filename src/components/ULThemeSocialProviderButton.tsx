import { Button, type ButtonProps } from "@/components/ui/button";
import { cn } from "@/lib/utils";

export interface ULThemeSocialProviderButtonProps
  extends React.HTMLAttributes<HTMLButtonElement>,
    ButtonProps {
  /**
   * The icon to be displayed in the social login button component
   */
  iconComponent: React.ReactNode | null;
  /**
   * The display Name to be displayed in the social login button component
   */
  displayName: string;
  /**
   * The button text to be displayed in the social login button component
   */
  buttonText: string;
  /**
   * Optional class names for additional styling or overriding default styles.
   */
  className?: string;
}

const ULThemeSocialProviderButton = ({
  onClick,
  variant = "outline",
  size = "default",
  iconComponent,
  displayName,
  buttonText,
  disabled = false,
  className,
  ...rest
}: ULThemeSocialProviderButtonProps) => {
  // Data Test Id needed for targetting element
  const dataTestId = `social-provider-button-${displayName.toLowerCase().replace(/\s+/g, "-")}`;

  // Base Styles getting applied for look and feel
  const baseStyles =
    "flex items-center justify-start w-full max-w-[320px] h-[52px] py-[14px] px-[16px] gap-x-4";

  // Keeping the button enable and disable styles as is. Will need to revisit this in future if needed
  const enabledStyles =
    "bg-white border-gray-mid text-text-default cursor-pointer";
  const disabledStyles =
    "bg-gray-mid/10 border-gray-mid/50 text-text-secondary cursor-not-allowed";

  // UL theme overrrides
  const variantThemeOverrides = {
    outline: cn(
      "theme-universal:font-button", //font-weight
      "theme-universal:rounded-button", // border-radius
      "theme-universal:border-(length:--ul-theme-border-button-border-weight)", // border-weight or border-width
      "theme-universal:border-(--color-secondary-button-border)", // border-color
      "theme-universal:text-(--color-secondary-button-label)", // text label color
      "theme-universal:hover:shadow-[var(--button-hover-shadow)]", // box-shadow on hover
      "theme-universal:focus:outline-none theme-universal:focus:ring-4 theme-universal:focus:ring-base-focus/15 theme-universal:focus:bg-base-focus/15" // focus base color
    ),
    primary: "", // Add primary overrides if needed
    secondary: "", // Add secondary overrides if needed
    destructive: "", // Add destructive overrides if needed
    ghost: "", // Add ghost overrides if needed
    link: "", // Add link overrides if needed
  };

  // Combine all theme classes with proper type safety
  const themeClasses = cn(variant && variantThemeOverrides[variant]);

  return (
    <Button
      variant={variant}
      size={size}
      className={cn(
        baseStyles,
        disabled ? disabledStyles : enabledStyles,
        themeClasses,
        className
      )}
      data-testid={dataTestId}
      disabled={disabled}
      onClick={onClick}
      {...rest}
    >
      {iconComponent && (
        <span className="w-5 h-5 flex items-center justify-center">
          {iconComponent}
        </span>
      )}
      <span className="overflow-hidden whitespace-nowrap text-ellipsis text-base">
        {buttonText}
      </span>
    </Button>
  );
};

export default ULThemeSocialProviderButton;
