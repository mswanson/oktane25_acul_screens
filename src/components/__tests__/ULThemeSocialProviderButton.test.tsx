import { fireEvent, render, screen } from "@testing-library/react";

import ULThemeSocialProviderButton from "../ULThemeSocialProviderButton";

describe("ULThemeSocialProviderButton Component", () => {
  const mockOnClick = jest.fn();

  const defaultProps = {
    iconComponent: <span data-testid="icon">Icon</span>,
    displayName: "Google",
    buttonText: "Sign in with Google",
    onClick: mockOnClick,
    disabled: false,
    className: "custom-class",
  };

  it("renders correctly and matches snapshot", () => {
    const { container } = render(
      <ULThemeSocialProviderButton {...defaultProps} />
    );
    expect(container).toMatchSnapshot();
  });

  it("displays the correct button text", () => {
    render(<ULThemeSocialProviderButton {...defaultProps} />);
    expect(screen.getByText("Sign in with Google")).toBeInTheDocument();
  });

  it("renders the correct icon", () => {
    render(<ULThemeSocialProviderButton {...defaultProps} />);
    expect(screen.getByTestId("icon")).toBeInTheDocument();
  });

  it("calls the onClick handler when clicked", () => {
    render(<ULThemeSocialProviderButton {...defaultProps} />);
    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it("disables the button when the disabled prop is true", () => {
    render(<ULThemeSocialProviderButton {...defaultProps} disabled={true} />);
    const button = screen.getByRole("button");
    expect(button).toBeDisabled();
  });

  it("applies additional class names", () => {
    render(<ULThemeSocialProviderButton {...defaultProps} />);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("custom-class");
  });

  it("renders the correct data-testid based on displayName", () => {
    render(<ULThemeSocialProviderButton {...defaultProps} />);
    const button = screen.getByTestId("social-provider-button-google");
    expect(button).toBeInTheDocument();
  });
});
