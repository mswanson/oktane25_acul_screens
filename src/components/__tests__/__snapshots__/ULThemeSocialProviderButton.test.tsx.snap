// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ULThemeSocialProviderButton Component renders correctly and matches snapshot 1`] = `
<div>
  <button
    class="focus-visible:ring-ring aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive theme-default:active:scale-[0.99] relative box-border shrink-0 gap-2 overflow-hidden text-sm font-medium whitespace-nowrap transition-all duration-150 ease-in-out outline-none focus-visible:ring-4 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 dark:bg-muted/50 hover:text-accent-foreground shadow-button-outlined-resting hover:shadow-button-outlined-hover hover:border-accent hover:bg-muted theme-default:before:from-primary/5 theme-default:before:to-primary/0 theme-default:before:absolute theme-default:before:top-0 theme-default:before:left-0 theme-default:before:block theme-default:before:h-full theme-default:before:w-full theme-default:before:bg-gradient-to-t theme-default:before:content-[''] border rounded-2xl has-[>svg]:px-3 flex items-center justify-start w-full max-w-[320px] h-[52px] py-[14px] px-[16px] gap-x-4 bg-white border-gray-mid text-text-default cursor-pointer theme-universal:font-button theme-universal:rounded-button theme-universal:border-(length:--ul-theme-border-button-border-weight) theme-universal:border-(--color-secondary-button-border) theme-universal:text-(--color-secondary-button-label) theme-universal:hover:shadow-[var(--button-hover-shadow)] theme-universal:focus:outline-none theme-universal:focus:ring-4 theme-universal:focus:ring-base-focus/15 theme-universal:focus:bg-base-focus/15 custom-class"
    data-testid="social-provider-button-google"
  >
    <span
      class="w-5 h-5 flex items-center justify-center"
    >
      <span
        data-testid="icon"
      >
        Icon
      </span>
    </span>
    <span
      class="overflow-hidden whitespace-nowrap text-ellipsis text-base"
    >
      Sign in with Google
    </span>
  </button>
</div>
`;
